### ISSUE TITLE
Shellcheck incorrectly complains about echo and printf commands

### ISSUE THREAD

**Issue by bvanassche** [koalaman/shellcheck#2663]
Rule id: SC2320
Version: https://github.com/koalaman/shellcheck/releases/download/stable/shellcheck-stable.linux.x86_64.tar.xz from 2023-01-11.
- This is documented on https://www.shellcheck.net/wiki/SC2320. However, I still think that this should be fixed.
- [X] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### Here's a snippet or screenshot that shows the problem:

```
git clone https://github.com/osandov/blktests.git
make check
```

#### Here's what shellcheck currently says:
```
2023-01-12T00:02:31.9746941Z common/multipath-over-rdma:134:24: warning: This $? refers to a condition, not a command. Assign to a variable to avoid it being overwritten. [SC2319]
2023-01-12T00:02:31.9747763Z common/null_blk:47:39: warning: This $? refers to echo/printf, not a previous command. Assign to variable to avoid it being overwritten. [SC2320]
2023-01-12T00:02:31.9748468Z common/null_blk:54:43: warning: This $? refers to echo/printf, not a previous command. Assign to variable to avoid it being overwritten. [SC2320]
2023-01-12T00:02:31.9749603Z tests/nvmeof-mp/rc:75:42: warning: This $? refers to echo/printf, not a previous command. Assign to variable to avoid it being overwritten. [SC2320]
2023-01-12T00:02:31.9750371Z tests/nvmeof-mp/rc:118:26: warning: This $? refers to a condition, not a command. Assign to a variable to avoid it being overwritten. [SC2319]
2023-01-12T00:02:31.9751192Z tests/srp/rc:99:42: warning: This $? refers to echo/printf, not a previous command. Assign to variable to avoid it being overwritten. [SC2320]
2023-01-12T00:02:31.9751829Z tests/srp/rc:105:44: warning: This $? refers to echo/printf, not a previous command. Assign to variable to avoid it being overwritten. [SC2320]
2023-01-12T00:02:31.9752642Z tests/srp/rc:263:26: warning: This $? refers to a condition, not a command. Assign to a variable to avoid it being overwritten. [SC2319]
2023-01-12T00:02:31.9753540Z tests/srp/rc:523:12: warning: This $? refers to echo/printf, not a previous command. Assign to variable to avoid it being overwritten. [SC2320]
2023-01-12T00:02:31.9754272Z tests/zbd/rc:83:14: note: $/${} is unnecessary on arithmetic variables. [SC2004]
2023-01-12T00:02:31.9754769Z tests/zbd/rc:85:14: note: $/${} is unnecessary on arithmetic variables. [SC2004]
2023-01-12T00:02:31.9755426Z tests/zbd/rc:87:13: note: $/${} is unnecessary on arithmetic variables. [SC2004]
2023-01-12T00:02:31.9755908Z tests/zbd/rc:88:13: note: $/${} is unnecessary on arithmetic variables. [SC2004]
2023-01-12T00:02:31.9756542Z tests/zbd/rc:89:13: note: $/${} is unnecessary on arithmetic variables. [SC2004]
2023-01-12T00:02:31.9757016Z tests/zbd/rc:95:14: note: $/${} is unnecessary on arithmetic variables. [SC2004]
2023-01-12T00:02:31.9757656Z tests/zbd/rc:97:14: note: $/${} is unnecessary on arithmetic variables. [SC2004]
2023-01-12T00:02:31.9758197Z tests/zbd/rc:128:10: warning: This $? refers to echo/printf, not a previous command. Assign to variable to avoid it being overwritten. [SC2320]
2023-01-12T00:02:31.9758995Z tests/srp/014:57:26: warning: This $? refers to a condition, not a command. Assign to a variable to avoid it being overwritten. [SC2319]
2023-01-12T00:02:31.9759596Z tests/srp/014:69:26: warning: This $? refers to a condition, not a command. Assign to a variable to avoid it being overwritten. [SC2319]
```

#### Here's what I wanted or expected to see:

None of the above messages. I think all of the above messages are either not useful or indicate a bug in shellcheck.

**Comment by TriMoon** [comment]
Just like any warnings in any computer language, they are not errors...

If you're sure it works as intended you can place disable comments...
_It is just helping you to code better..._

**Comment by bvanassche** [comment]
Thanks for having taken the time to provide a detailed reply. However, it is not clear to me why `[ ... ] || return $?` is considered bad style? My view is that `[ ... ] || return $?` is idiomatic in shell code.

The URL of the blktests project is [https://github.com/osandov/blktests](https://github.com/osandov/blktests).

The blktests project has adopted a "zero shellcheck warnings" policy.

There are 152 occurrences of " || return $?" in the blktests project. Changing all occurrences would require a significant effort.

**Comment by austin987** [comment]
https://www.shellcheck.net/wiki/SC2320

**Comment by port19x** [comment]
> Thanks for having taken the time to provide a detailed reply. However, it is not clear to me why `[ ... ] || return $?` is considered bad style? My view is that `[ ... ] || return $?` is idiomatic in shell code.
> 
> The URL of the blktests project is https://github.com/osandov/blktests.
> 
> The blktests project has adopted a "zero shellcheck warnings" policy.
> 
> There are 152 occurrences of " || return $?" in the blktests project. Changing all occurrences would require a significant effort.

It's bad style because it references the previous result and introduces coupling.

If we have some code, as in the wiki page, like the following
```sh
mycommand
echo "Command exited with $?"
```

And we now, in the process of development or refactoring, switch up our structure
```sh
mycommand
myfailingcommand
echo "Command exited with $?"
```

We get unexpeced behaviour, where as with something more idiomatic like
```sh
mycommand
A=$?
myfailingcommand
echo "Command exited with $A"
```
We at least somewhat decouple the initial command from the next check.

It's kinda bad either way. `||` and `&&` exist.
But if you have to do it, don't shoot yourself in the foot any harder than you have to.

I initially confused this with [SC2312](https://www.shellcheck.net/wiki/SC2312), which may be more enlightening to you

**Comment by TriMoon** [comment]
@bvanassche 
> My view is that `[ ... ] || return $?` is idiomatic in shell code.

Do you actually understand what it does, step-by-step? :thinking: 
W all understand what the human means here, but how does `bash` process that line?
If you understand THAT you will understand that the `$?` used here is completely  "bad-code"...
It should use `return ###` (xxx being any number, 0 <= xxx <= 255), but never `$?`.

> Changing all occurrences would require a significant effort.

Fixing bad code always requires efforts, the magnitude is just related to the amount of bad coding in the first place.
But it makes the code-base much better and prevents bugs that are hard, to impossible, to find if not fixed.

People think that writing shorter (cryptic) code makes it faster, which is just not true. outside of direct assembler code...
Interpreted languages convert the code in the same final code for the CPU no matter if you use cryptic code like in the example we talk about or use the longer 3  line version...
eg: this is same for the final code provided to the CPU:
```
if test ....; then
  return xxx
fi
```

**Comment by oliv3r** [comment]
With SC2312, I wonder (probably best to file a new issue).
```sh
printf "$(cat '/etc/hostname')"
```
will yield the dreaded SC2312 error. Probably rightfully so. But in this particular case (as with most of these) I think that shellcheck should accept not just `|| true` to ignore the warning (or accept the fact), but as the rationale/exception says 'if you have other means.

While this will be a very slippery slope, I think
```sh
printf "$(cat '/etc/hostname' || echo 'Unknown')"
# and
printf "$(cat '/etc/hostname' || printf 'Unknown')"
```
I think are both valid choices in addition to `|| true`. I don't think shellcheck fails on `printf "$(echo 'test')"` does it? As echo and friends are considered to be okay to ignore return values (I'm not even sure lol).

So for SC2312 specifically, I would argue the above?

**Comment by Alexander-Shukaev** [comment]
> [@bvanassche](https://github.com/bvanassche)
> 
> > My view is that `[ ... ] || return $?` is idiomatic in shell code.
> 
> Do you actually understand what it does, step-by-step? 🤔 W all understand what the human means here, but how does `bash` process that line? If you understand THAT you will understand that the `$?` used here is completely "bad-code"... It should use `return ###` (xxx being any number, 0 <= xxx <= 255), but never `$?`.

Cool down bro, this code is absolutely legit.  Nobody has to devise a special return code for this case unless it's really needed...  The rest of your post is just a opinionated rant.

**Comment by TriMoon** [comment]
> Cool down bro, this code is absolutely legit. Nobody has to devise a special return code for this case unless it's really needed... The rest of your post is just an opinionated rant.

Sorry about that, i must have been in a bad mood indeed...
I didn't intend to write it with such a hard tone, was meant as educational...
Again excuses for the tone of the text...
